packages/formula-editor/src/parser.ts

```diff
@@ -0,0 +1,465 @@
+import Big from "big.js";
+import { Expectation, Queue, Stack } from "./helpers.js";
+import { Recommender } from "./recommendor.js";
+
+export interface ParseResult {
+  recommendations: string[] | null;
+  formattedContent: HTMLBodyElement | null;
+  formattedString: string | null;
+  newCursorPosition: number;
+  errorString: string | null;
+}
+
+export interface CalculateResult {
+  result: number | undefined;
+  errorString: string | null;
+}
+
+export class Parser {
+  constructor(variables: Map<string, number>, minSuggestionLen: number) {
+    this.variables = variables;
+
+    this._recommender = new Recommender(this.variables, minSuggestionLen);
+  }
+
+  private _recommender: Recommender;
+
+  variables: Map<string, number>;
+  mathematicalOperators: Set<string> = new Set(["^", "+", "-", "*", "/"]);
+  operatorPrecedence: { [key: string]: number } = {
+    "^": 3,
+    "/": 2,
+    "*": 2,
+    "+": 1,
+    "-": 1,
+  };
+
+  parseInput(
+    formula: string,
+    prevCurPos: number | null = null,
+    recommendation: string | null = null
+  ): ParseResult {
+    let tokens = formula.split(/([-+(),*^/:?\s])/g);
+
+    // Stores the positions of opening parentheses. This allows us to
+    // show "Unclosed parenthesis error" for positions which are far behind
+    // our current token
+    let parentheses = new Stack<number>();
+
+    // The HTML formatted string which we eventually show on the view.
+    let formattedString = ``;
+
+    // The expectation that we have for the current token.
+    let expectation = Expectation.VARIABLE;
+
+    // Position of the current token in the formula string.
+    let currentPosition = 0;
+
+    // Previous 'token' (not a space or a new line) that we just encountered.
+    let previousToken = "";
+
+    // The object that we return as the output of the parsing result.
+    let parseOutput: ParseResult = {
+      recommendations: null,
+      formattedContent: null,
+      formattedString: null,
+      newCursorPosition: prevCurPos ?? -1,
+      errorString: null,
+    };
+
+    console.log(tokens);
+
+    tokens.forEach((token) => {
+      // It is a number is either it's in the defined variables, or
+      // it's a valid number literal.
+      let isNumber = this.variables.has(token) || !Number.isNaN(Number(token)),
```

Comment:
how about using a `tokenType=number | operator | bracket | space`
instead of having 4 such variables.

Reply:
Actually I was thinking of having an Enum for this. Seemed like a more structured approach to me.